### DISCUSSION TITLE
Rig Optimization

### DISCUSSION THREAD

**Discussion by jjburton** [jjburton/cgmTools#208] [Ideas]
I've been following some optimization ideas in some td groups and I'm wondering if any of you all are interested in helping r&d some of the things. If so, lemme know.

One cycle issue for parallel seems to be the common practice (as we do) of having the settings (fk/ik swtich etc) follow the rig. It creates a loop that slows down eval. I've got a few ideas here I'd like some help experimenting with.

### Take Aways
If you're seeing speed issues. Here's some things to try. Most of our builds are game assets and playback at speed fine. However some of our users are doing much higher res render projects and there are some definite places to pick up frames.

- **Control Curves** | turn off the p for playback. Drawing these can take 5-6fps off your playback on higher res rigs
- **RigBlocks** | Delete these from your rig files. I keep them in a file I call builds and then rigs don't have them. Because these are tagged as Controllers to enable pickwalking, they can get pulled into evaluation depending on your playback options
- **Reference Frames** | having a frame sequence playing in your scenes eats up your playback speed

**Comment by jjburton** [comment]
### Looking at someone's file
- Image sequence camera in scene is a 4-5fps hit
- Proxy is faster than lots of geo clusters.
- Might be worth a version with a unified mesh. Need to test

Game rig performance is pretty good but the higher tier render work needs some work. 

@benngarnish | chime in here with any thoughts

- Spline IK on the tail might be worth a try on the build to see if the ribbon solve is a big hit. I have some other roll joint setup options as well we could try.

**Reply by benngarnish** [reply]
Good point on the geo clusters. 
Let me see what I can do on that front to see if performance increases. 
LOD'ing that combined geo would also be useful me thinks. 

Trying the other options you've suggested today like spline ik instead of ribbon.

**Reply by jjburton** [reply]
### Second pass looking at @benngarnish file
- The eye setup is using the closest point setup rather than the fan joint which would probably be faster. The closest point nodes are a hit and you have a LOT of lid joints. To test I just deleted the eye rig leaving the lid joints and found this faster. You can try the simpler fan setup or have a version of the rig asset with detailed face stuff and without
- The settings controls are a bottle neck with parallel because the are typically residing at the end of the chains while affecting the whole limb chains creating a cycle cluster. To test I parented all the settings to the world and again saw improvements
- Proxy geo is definitely faster than the skin cluster though not by a huge amount. There are a number of geo chunks that could probably be unified for a faster performance asset and kept separate in another render asset if that proves to be true.


I have more stuff to look at but I'm starting to see some definite areas where we can get some improvements just by what build options users are doing.

**Comment by benngarnish** [comment]
Proxy only shows 25fps right away
Switching on the MRS created geo from build goes to 9fps

**Comment by jjburton** [comment]
### Findings 06.23
I'll keep updating this.

- Image sequence cameras in scene are a 4-5fps hit
- Proxy geo is faster than skinning

Eye
- The fan setup will probably be faster

Segment
- SplineIK will probably be faster than the ribbon setup
- When you have  control joint count == joint count it should just be on spline

LIMB
- The settings at the end of the chain is a hit. Trying to think through how to resolve this best.

**Comment by jjburton** [comment]
### Findings 06.26

DynParentGroupSetup
This is a large chunk of slowdown. 
- Tried just killing the contraints and saw a nice jump in performance. This isn't going to work but is a sign that there are improvements to be had here
- Broke the live connection from the objects to the DynParentGroup. This saw similar benefits. Live connection may be the culprit
- Charles material said that the number of space options might be a hit here as well so optimization there might be in order

Rebuild curve nodes
- Removed x rebuild curve nodes to test performance


Result
- Base | 20fps | 73ms
- No Parent Bind | 20.4fps | 73ms ....this one is a wash for just the bind skeleton, need to look if other places could benefit
- Settings Out | 29fps | 49ms
- No Rebuild |  23fps |  68ms - 32 rebuild curve nodes
- No live dynParent | 20fps | 65ms



```python

for mObj in cgmMeta.asMeta(mc.ls('*_settings_anim')):
    mParent = mObj.getParent(asMeta=1)
    mParent.dagLock(0)
    mParent.p_parent = False
    
for mObj in cgmMeta.asMeta(mc.ls('*_settings_anim')):
    mRigNull = mObj.getMessageAsMeta('cgmOwner')
    if mRigNull:
        mRoot = mRigNull.getMessageAsMeta('rigRoot')
        if mRoot:
            mParent = mObj.getParent(asMeta=1)
            mParent.dagLock(0)
            mParent.p_parent = mRoot.dynParentGroup.p_parent
    

import cgm.core.lib.attribute_utils as ATTR
for mObj in r9Meta.getMetaNodes(mTypes = 'cgmDynParentGroup',nTypes=['transform']):
    mDynChild = mObj.dynChild
    for a in ['orientTo','space','follow']:
        if mDynChild.hasAttr(a):
            print mDynChild.mNode
            for p in ATTR.get_driven(mDynChild.mNode,a):
                print p
                ATTR.break_connection(p)

print len(mc.ls(type='rebuildCurve'))
mc.delete(mc.ls(type='rebuildCurve'))
```

**Comment by benngarnish** [comment]
Running the above code looses the connection on the settings so when I move my wing limb around the settings COG no longer follows. The same goes for the finger block settings COG too!

**Reply by jjburton** [reply]
Yes. It's parenting them to the world :) That's what I was talking about with how to show them in another way -perhaps in a curve ui that snaps to the camera.

**Reply by benngarnish** [reply]
Why on earth does this have issues with performance? So strange 'scratching head...'

**Reply by benngarnish** [reply]
Looks as if the Space options are broken too when you run the code above

**Reply by jjburton** [reply]
Yes the space one will break things. They're not meant to be actual rig patches. Just tests :) I'd like to know if you see similar results in other rigs.

The reason for the settings thing is a cycle cluster - the control affects the blend of ik to fk which informs where that setting control needs to be so it's cycling. Moving it out lets maya say, okay, this is the setting, this control doesn't get changed by the evaluation and it's a clean graph.

**Reply by benngarnish** [reply]
Would throwing that onto the shape node help?

**Reply by jjburton** [reply]
Don't think so but you're welcome to try.

**Comment by jjburton** [comment]
### Trying a test with benn's file
2018

First test - 
- Each settings control
- Created dup control to hold real values
- Moved settings attributes over to those
- Proxied them back to the root

Pre result
-732ms | ~14pfs
![image](https://user-images.githubusercontent.com/48218358/131229521-41c4cc35-56d3-4ade-88bd-b7995b9b70ad.png)

BUT reopening the file brought it about where the post one was. why that is I'm not sure

Post
-120ms | 17-19fps

Result
- Doesn't feel like it did much

------------------------------------------------------------------------
Test 2 - Settings to world
- Just parent the settings controls to the world
-
Result
- 105ms
- 16-17fps

**Comment by jjburton** [comment]
### Same tests in maya version
2019
- base | 18.6fps || cached 25
- proxy attr 1 | 18.6 || cached 25fps
- parent to world 16.6  || cached 25

2020 
- base 20fps || cached 25
- proxy attr 1 20fps || cached 25
- parent to world 20 || cached 25

**Comment by jjburton** [comment]
### Skinning on the bind skeleton moved to the rig skeleton

As this actually adds an extra constraint set per joint in your rig this is a potential area to investigate.

Need to write a skinning tranfer utility to move them from the bind to the rig joints


...Wrote this.

1. Repair rigJoint wiring on bind
2. Save the skinDat
3. Delete cluster
4. ApplySkin with skinDat

```python
import cgm.core.mrs.lib.post_utils as POST
reload(POST)

POST.repair_rigJointWiring()

import cgm.core.lib.skinDat as SKINDAT
mSkin = SKINDAT.data(sourceMesh = 'xxxxxxx_geo')
mSkin.updateSourceSkinData()
mSkin.write()
mSkin.validateTargetMesh('xxxxxxx_geo')
mSkin.applySkin(influenceMode = 'msgLink')
```

113ms | 19fps

**Comment by jjburton** [comment]
### DynParent attribute move via proxy
Similar to the settings attribute but do so with the dynParentAttributes

Finished this mockup as well. Not really seeing much. @benngarnish | it's in the testing set along with one I also deleted rigBlocks from.

**Comment by jjburton** [comment]
### Color some nodes for profile work
Charles recommended maybe trying this...raw code, will need to incorporate to builds.

```python
from maya import cmds as mc
import pymel.core as pm

"""
This is just a python wrapper based on the code here:
https://knowledge.autodesk.com/support/maya/learn-explore/caas/CloudHelp/cloudhelp/2020/ENU/Maya-Customizing/files/GUID-8D5FFC12-608C-45EA-B035-1AB56F3C42F1-htm.html

The issue with the original is that it doesn't do creation-time checks.  This code also loops
over each object in turn to let you set the same tag to a variety of selected nodes in one go.
"""

def add_profiler_tag(tag_value, color):
	"""
	Set the profiler tag value and color for all selected nodes.
	Note that this process *requires* looping over all nodes one by one.

	:param tag_value: Tag name to give the selected nodes
	:type tag_value: str
	:param color: RGB color array, colors from 0 to 255
	:type color: List[int]
	"""
	sel = mc.ls(sl=True)
	assert len(sel), "add_profiler_tag requires a node selection."
	
	## create the structure if it doesn't exist already
	if not "NodeProfileStruct" in mc.dataStructure(q=True):
		mc.dataStructure(format="raw", asString="name=NodeProfileStruct:string=NodeProfileTag:int32=NodeProfileTagColor")

	for node in sel:
		mc.select(node, r=True)
		## addMetadata attaches space for the metadata on the selected nodes
		## notice that this is called once for each part of the structure defined above!
		extant_metadata = mc.addMetadata(q=True, channelName=True)
		if not "ProfileTag" in extant_metadata:
			mc.addMetadata(streamName="ProfileTagStream", channelName="ProfileTag", structure="NodeProfileStruct")
		if not "ProfileTagColor" in extant_metadata:
			mc.addMetadata(streamName="ProfileTagColorStream", channelName="ProfileTagColor", structure="NodeProfileStruct")
		
		## now that the struct exists and we've attached these streams to the 
		## selected object, we can fill in the streams with the info we want
		mc.editMetadata(streamName="ProfileTagStream", memberName="NodeProfileTag", channelName="ProfileTag", stringValue=tag_value, index=0)
		
		mc.editMetadata(streamName="ProfileTagColorStream", memberName="NodeProfileTagColor", channelName="ProfileTagColor", value=color[0], index=0)
		mc.editMetadata(streamName="ProfileTagColorStream", memberName="NodeProfileTagColor", channelName="ProfileTagColor", value=color[1], index=1)
		mc.editMetadata(streamName="ProfileTagColorStream", memberName="NodeProfileTagColor", channelName="ProfileTagColor", value=color[2], index=2)

	mc.select(sel, r=True)
```
